constraint/pkg/client/client.go

```diff
@@ -135,29 +167,72 @@ func (c *Client) AddTemplate(ctx context.Context, templ *templates.ConstraintTem
 		return resp, err
 	}
 
-	if err := c.driver.AddTemplate(ctx, templ); err != nil {
+	newDriverN := c.driverForTemplate(templ)
+
+	driver, ok := c.drivers[newDriverN]
+	if !ok {
+		return resp, fmt.Errorf("%w: available drivers: %v, wanted %q", clienterrors.ErrNoDriver, c.driverPriority, c.driverForTemplate(templ))
+	}
+
+	// TODO: because different targets may have different code sets,
+	// the driver should be told which targets to load code for.
+	// this is moot right now, since templates only have one target
+	if err := driver.AddTemplate(ctx, templ); err != nil {
 		return resp, err
 	}
 
 	templateName := templ.GetName()
 
-	template := c.templates[templateName]
+	cacheEntry := c.templates[templateName]
 
 	// We don't want to use the usual "if found/ok" idiom here - if the value
 	// stored for templateName is nil, we need to update it to be non-nil to avoid
 	// a panic.
-	if template == nil {
-		template = &templateClient{
-			constraints: make(map[string]*constraintClient),
+	if cacheEntry == nil {
+		cacheEntry = newTemplateClient()
+		c.templates[templateName] = cacheEntry
+	}
+
+	cacheEntry.activeDrivers[newDriverN] = true
```

Comment:
is activeDrivers thread safe? does it need any locks?

Reply:
It would not be thread safe, but I believe `cacheEntry` is gated by the `c.mtx` lock above?